src/main/java/org/springframework/integration/kafka/listener/ConcurrentMessageListenerDispatcher.java

```diff
@@ -0,0 +1,173 @@
+		Assert.notNull(delegateListener, "A delegate must be provided");
+		this.delegateListener = delegateListener;
+		this.partitions = partitions;
+		this.consumers = consumers;
+		this.offsetManager = offsetManager;
+	}
+
+	public ErrorHandler getErrorHandler() {
+		return errorHandler;
+	}
+
+	public void setErrorHandler(ErrorHandler errorHandler) {
+		this.errorHandler = errorHandler;
+	}
+
+	public OffsetManager getOffsetManager() {
+		return offsetManager;
+	}
+
+	public void setOffsetManager(OffsetManager offsetManager) {
+		this.offsetManager = offsetManager;
+	}
+
+	public int getQueueSize() {
+		return queueSize;
+	}
+
+	public void setQueueSize(int queueSize) {
+		this.queueSize = queueSize;
+	}
+
+	@Override
+	public void start() {
+		synchronized (lifecycleMonitor) {
+			if (!isRunning()) {
+				initializeAndStartDispatching();
+				this.running = true;
+			}
+		}
+	}
+
+	@Override
+	public void stop() {
+		synchronized (lifecycleMonitor) {
+			if (isRunning()) {
+				this.running = false;
+				delegates.flip().keyBag().toSet().forEach(stopDelegateProcedure);
+			}
+		}
+	}
+
+	@Override
+	public boolean isRunning() {
+		return running;
+	}
+
+	public void dispatch(KafkaMessage message) {
+		delegates.get(message.getMetadata().getPartition()).enqueue(message);
+	}
+
+	private void initializeAndStartDispatching() {
+		// allocate delegate instances index them
+		List<QueueingMessageListenerInvoker> delegateList = new ArrayList<QueueingMessageListenerInvoker>(consumers);
+		for (int i = 0; i < consumers; i++) {
+			QueueingMessageListenerInvoker blockingQueueMessageListenerInvoker = new QueueingMessageListenerInvoker(queueSize, offsetManager, delegateListener);
+			if (errorHandler != null) {
+				blockingQueueMessageListenerInvoker.setErrorHandler(errorHandler);
+			}
+			delegateList.add(blockingQueueMessageListenerInvoker);
+		}
+		// evenly distribute partitions across delegates
+		delegates = Maps.mutable.of();
+		int i = 0;
+		for (Partition partition : partitions) {
+			delegates.put(partition, delegateList.get((i++) % consumers));
+		}
+		// initialize task executor
+		if (this.taskExecutor == null) {
+			this.taskExecutor = Executors.newFixedThreadPool(consumers, THREAD_FACTORY);
+		}
+		// start dispatchers
+		delegates.flip().keyBag().toSet().forEachWith(startDelegateProcedure, taskExecutor);
+	}
+
+	@SuppressWarnings("serial")
+	private static class StopDelegateProcedure implements Procedure<QueueingMessageListenerInvoker> {
+		@Override
+		public void value(QueueingMessageListenerInvoker delegate) {
+			delegate.stop();
+		}
+	}
+
+	@SuppressWarnings("serial")
+	private static class StartDelegateProcedure implements Procedure2<QueueingMessageListenerInvoker, Executor> {
+		@Override
+		public void value(QueueingMessageListenerInvoker delegate, Executor executor) {
+			delegate.start();
+			executor.execute(delegate);
```

Comment:
No, I don't mind of those field for these functions/procedures, because, yes: we must avoid any `new...` as much as possible. I want to pay your attention that we can avoid here `gs-collections` altogether. 
From other side we can avoid extracting these classes but just provide their inner impl from the declaration:
````java
private static final Procedure2<QueueingMessageListenerInvoker, Executor> startDelegateProcedure =
			new Procedure2<QueueingMessageListenerInvoker, Executor>() {

				@Override
				public void value(QueueingMessageListenerInvoker delegate, Executor executor) {
					delegate.start();
					executor.execute(delegate);
				}

			};
````